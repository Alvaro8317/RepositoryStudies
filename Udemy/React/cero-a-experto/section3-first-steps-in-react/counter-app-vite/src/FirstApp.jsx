/*
 * Toda app de react empieza con un componente o functional component, como un   punto de entrada
 * Las clases no son recomendadas en react, mejor con componentes basados en funciones o functional components
 * Todos los nombres de los componentes van con PascalCase
 * El funcional component no va usualmente en main.jsx
 * Equivalente a document.createElement
 * No es recomendable porque title debe de venir pero se añade una lógica al componente innecesaria, aquí entra los propTypes
 * Escribir el functional component.propTypes sirve para indicar que tipo de dato se requiere y si es obligatorio o no, si se envía un datatype incorrecto renderiza pero deja el error en consola. Con isRequired se indica si es obligatorio o no un campo
 * rafc genera un nuevo componente con las extensiones de vscode
 * Se pueden crear los valores por defecto en defaultProps o en los parámetros del functional component, pero mejor práctica en defaultProps porque con 5 parámetros por defecto se volvería difícil de leer el código
 */
import PropTypes from 'prop-types';
const App = ({ title = 'No hay título', subtitle = 'No hay subtitulo' }) => {
  // if (!title) {
  //   throw new Error('El title no existe');
  // }
  return (
    <>
      <div>Hola React</div>
      <h1 data-testid='test-title'>{title}</h1>
      <h2>{subtitle}</h2>
    </>
  );
};
/* Tipos de dato y si son requeridos o no */
App.propTypes = {
  title: PropTypes.string.isRequired,
  subtitle: PropTypes.string.isRequired,
};
/* Props por defecto */
// App.defaultProps = {
//   subtitle: 'No hay subtitulo',
// };
export default App;
