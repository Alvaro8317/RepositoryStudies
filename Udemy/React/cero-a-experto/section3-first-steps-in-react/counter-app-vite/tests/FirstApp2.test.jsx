const { render, screen } = require('@testing-library/react');
import FirstApp from '../src/FirstApp';
/*
 * Se crea este archivo para emplear mejores prácticas
 * Screen es el objeto que se está renderizando
 */
describe('Test of First App Component', () => {
  const title = 'Hola, soy un título';
  const subtitle = 'Hola, soy un subtítulo';
  test('Should match the snapshot', () => {
    const { container } = render(<FirstApp title={title} />);
    expect(container).toMatchSnapshot();
  });
  test('should show the title "Hola, soy un título"', () => {
    render(<FirstApp title={title} />);
    // screen.debug();
    expect(screen.getByText(title)).toBeTruthy();
  });
  test('should show the title in h1', () => {
    render(<FirstApp title={title} />);
    expect(screen.getByRole('heading', { level: 1 }).innerHTML).toContain(
      title
    );
  });
  test('should show the subtitle by props', () => {
    render(<FirstApp title={title} subtitle={subtitle} />);
    expect(screen.getAllByText(subtitle).length).toBe(1);
  });
});
