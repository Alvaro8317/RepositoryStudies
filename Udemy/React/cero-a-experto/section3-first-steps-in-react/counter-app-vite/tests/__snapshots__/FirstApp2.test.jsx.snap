// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test of First App Component Should match the snapshot 1`] = `
<div>
  <div>
    Hola React
  </div>
  <h1
    data-testid="test-title"
  >
    Hola, soy un t√≠tulo
  </h1>
  <h2>
    No hay subtitulo
  </h2>
</div>
`;
